<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2018-02-02 16:35:57 -0600">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    
    <link rel="shortcut icon" type="image/x-icon" href="/SELU.gif" />
    
    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->
    <title>Systematics Lab: Multiple Sequence Alignment</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      
      <a href="http://datacarpentry.org" class="pull-left">
        <img class="navbar-logo" src="../assets/img/SELU.gif" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	
        <li><a href="../conduct/">Code of Conduct</a></li>

	
        
        <li><a href="../setup/">Setup</a></li>
        <li><a href="../reference/">Reference</a></li>
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Episodes <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            <li><a href="../00-LONI/">LONI Login Activity</a></li>
            
            <li><a href="../01-UNIX/">Downloading the class files</a></li>
            
            <li><a href="../02-UNIX/">Introducing the Shell</a></li>
            
            <li><a href="../03-Files/">Navigating Files and Directories</a></li>
            
            <li><a href="../04-Dirs/">Working With Files and Directories</a></li>
            
            <li><a href="../05-Pipes/">Pipes and Filters</a></li>
            
            <li><a href="../06-Loops/">Loops</a></li>
            
            <li><a href="../07-Script/">Shell Scripts</a></li>
            
            <li><a href="../08-ProjectOrganization/">Project Organization</a></li>
            
            <li><a href="../09-GitIntro/">Setting Up Git</a></li>
            
            <li><a href="../10-GitSetUp/">Setting Up Git</a></li>
            
            <li><a href="../11-GitRepositories/">Creating a Repository</a></li>
            
            <li><a href="../12-GitChanges/">Tracking Changes</a></li>
            
            <li><a href="../13-Github/">Remotes in GitHub</a></li>
            
            <li><a href="../14-MSA/">Multiple Sequence Alignment</a></li>
            
          </ul>
        </li>
	

	
	
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
          </ul>
        </li>
	

	
        <li><a href="../license/">License</a></li>
      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>





<div class="row">
  <div class="col-md-1">
    <h3>
      
      <a href="../13-Github/"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-md-10">
    
    <h3 class="maintitle"><a href="../">Systematics Lab</a></h3>
    <h1 class="maintitle">Multiple Sequence Alignment</h1>
    
  </div>
  <div class="col-md-1">
    <h3>
      
      <a href="../"><span class="glyphicon glyphicon-menu-up" aria-hidden="true"></span><span class="sr-only">lesson home</span></a>
      
    </h3>
  </div>
</div>


<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 30 min
      <br/>
      <strong>Exercises:</strong> 0 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How do I assign homology in molecular data sets?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Explain how multiple sequence alignment works.</p>
</li>
	
	<li><p>Explain how guide-tree approaches perform multiple sequence alignment.</p>
</li>
	
	<li><p>Explain how multiple genes can be efficiently indivdually aligned.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p>##Overview of multiple sequence alignment</p>

<p>As you have discussed in class, building a phylogeny is dependent on homology.
If we are using character data, typically an expert has assigned homology based on their knowledge of the fossil record, developmental sequence and other sources of information.</p>

<p>When we are working with molecular data, this is not possible. The goal of multiple sequence alignment (MSA) is to introduce gaps such that each residue in a <em>column</em> will be homologous. At the end of this process, we have a data matrix in which the residues in each column are hypothesized to be descended from a residue in a common ancestor.</p>

<p>This typically takes the form of several possible moves:</p>
<ol>
  <li>Rewarding matches.
    <ul>
      <li>Add to alignment score</li>
    </ul>
  </li>
  <li>Penalize opening gaps.
    <ul>
      <li>Decrement score. Typically 5 points to open.</li>
    </ul>
  </li>
  <li>Penalize rare substitutions, but not slight differences.
    <ul>
      <li>Decrement score.</li>
    </ul>
  </li>
  <li>Penalize extending gaps.</li>
</ol>

<p>The alignment that maximizes the total score is considered to be the best. There are multiple ways of achieving an alignment. We will cover three of them today.</p>

<p>Exercise One &amp; Two on board</p>

<p>We often use what are called <em>scoring matrices</em> to obtain the score of individual substitutions. We will talk about many types of substitution matrices. One that is commonly used in alignment is BLOSUM 62, which is based on empirically-calculated likelihoods of seeing different types of substitutions.</p>

<p><img src="../fig/blosum62.png" alt="Blosum" /></p>

<p>Exercise 3 and 4 on board.</p>

<p>As we can see, involving true biological knowledge to score alignments can improve alignment scores, sometimes dramatically!</p>

<p>The three exercises we will do today will focus on different strategies that may be deployed to find the optimal alignment.</p>

<p>##Progressive MSA</p>

<p>Pairwise Multiple Sequence Alignment does exactly what we discussed above, for all your sequences. Typically, this is performed by first performing a pairwise MSA, as we did above, between the two sequences with the least differences. Then, sequences are added in the order of least differences. Progressive alignments score different alignment possibilities based on the alignment and user-specified scoring matrices, such as BLOSUM 62, or others. Clustal is an example of progressive alignment, but is known to be fairly innaccurate. Below, we use T-COFFEE to get an alignment.</p>

<p>First, we will download and compile T_Coffee</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>git clone git@github.com:cbcrg/tcoffee.git tcoffee
cd tcoffee
cd compile
make t_coffee
</code></pre></div></div>
<p>We haven’t played with any software yet in this class, so I’ll now have you create a single location for all your software to live. Change directories into your home. Create a directory called ‘bin’. Move the t-coffee executeable into it.</p>

<p>Now, move into the SELUSys repository. Enter the data directory, and the MSALab directory. We’ll now run t_coffee on its default settings (a BLOSUM62 matrix with no gap penalty):</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>t_coffee data/sh3.fasta
</code></pre></div></div>
<p>Now, we’ll try a gap-opening penalty:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>t_coffee data/sh3.fasta -matrix blosum62mt -gapopen 5 -outfile=gapopen5
</code></pre></div></div>

<p>And one with a gap-extending penalty:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>t_coffee data/sh3.fasta -matrix blosum62mt -gapopen 5 gapextend 5 outfile gapopen5
</code></pre></div></div>

<p>Try one or two more additional alignments, such as increasing the gap open or extension.</p>

<p>Now, transfer your .aln files to your personal machine using the scp command. Use <a href="http://wasabiapp.org/">Wasabi</a>to view them.</p>

<p>##Iterative Approaches to MSA</p>

<p>Progressive aligners make a set of assumptions, and apply those assumptions to the whole set of sequences across a phylogeny. But many datasets are fairly large, and may have different evolutionary dynamics across the tree. <em>Iterative aligners</em> use tree information to guide the process of making the alignment.</p>

<p><img src="../fig/pasta.jpg" alt="Diagram of Sate/PASTA Algorithm" /></p>

<p>Now, we will download PASTA:</p>

<pre><code class="language-unix">git clone https://github.com/smirarab/sate-tools-linux.git
git clone https://github.com/smirarab/pasta.git
</code></pre>

<p>And we will load a couple libraries required by pasta:</p>
<pre><code class="language-unix">module load java
module load python/3.5.2-anaconda-tensorflow
</code></pre>

<p>Now, we build pasta:</p>

<pre><code class="language-unix">python setup.py develop --user
</code></pre>

<p>Before you leave the directory, give yourself permission to execute this file.</p>

<p>We will be submitting these jobs to the nodes of the LONI computer. Change directories back into your home directory.</p>

<p>Now, change directories into your work directory. Create a directory called “lab1”, containing three subdirectories: “data”, “scripts”, and “output”.</p>

<p>Enter the SELUSys2018 class directory. In it, there is a directory called scripts. List this directory. If you have a script called “pasta_script1”, copy this script into your lab1 scripts directory. If you do not have this file, type</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>git pull
</code></pre></div></div>

<p>to update this directory. Move the script into your lab1 scripts directory.</p>

<p>Use a text editor to open pasta_script1. Let’s unpack this script together. Feel free to make notes in your script about what different parts mean.</p>

<p>Now, we will run the script. First:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qsub -I -V -A loni_selu_sys -q single -l nodes=1:ppn=1,walltime=3:00:00
</code></pre></div></div>

<p>This will place us on a node so we can use lots of memory for this job.</p>

<p>Execute the actual pasta run:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>python ~/pasta/run_pasta.py -i data/pythonidae.fasta
</code></pre></div></div>

<p>As this runs, we will discuss the output that is appearing to the screen.</p>

<p>Now, we will look at the output. Create an “exercise_one” directory in the output directory. Move all of the outputs into the exercise_one directory.</p>

<p>Copy the pastajob.marker001.pythonidae.aln and pastajob_temp_iteration_2_tree.tre into your user home:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cp output/exercise_one/pastajob_temp_iteration_2_tree.tre ~/
cp output/exercise_one/pastajob.marker001.pythonidae.aln ~/
</code></pre></div></div>

<p>Now, we will copy these files to our local machine. Open a new terminal window. Navigate to the class repository, and make a lab1 directory, with somewhere to store the output. Change directories into it. Now, we will use secure copy to move the alignment file into our directory:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>scp &lt;username&gt;@qb2.loni.org:~/pastajob.marker001.pythonidae.aln .
scp &lt;username&gt;@qb2.loni.org:~/pastajob_temp_iteration_2_tree.tre .
</code></pre></div></div>

<h3 id="choice-of-tree-estimator">Choice of Tree Estimator</h3>

<p>Pasta allows us to build a tree for iterative estimation in different ways. <a href="http://www.microbesonline.org/fasttree/">FastTree</a> has some known <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0027731">accuracy issues</a>. That’s probably not an issue for an early stage analysis that will be improved. But let’s have a look at what happens if we use a better alignment.</p>

<p>In the class repository, there is a second script for today, pasta_script2. Copy it into your lab1 folder. Open it in nano. We will need to edit the working directory, and your email address.</p>

<p>From the lab1 directory, submit the script to the nodes using:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qsub pasta_script2
</code></pre></div></div>

<p>You can check on the progress of your job by typing</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qstat -u &lt;your username&gt;
</code></pre></div></div>

<p>You’ll get output something like this:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Req'd   Elap
Job ID               Username    Queue    Jobname          SessID NDS   TSK    Memory Time  S Time
-------------------- ----------- -------- ---------------- ------ ----- ------ ------ ----- - -----
446904.qb3           amwright    workq    binarysims        65908     1     20    --  36:00 R 29:17
447323.qb3           amwright    workq    MSsims           101460     1     20    --  36:00 R 06:59
447478.qb3           amwright    single   pasta             99242     1      1    --  01:00 R 00:01
</code></pre></div></div>

<p>Unfortunately, the way this cluster is configured, you won’t see any output from the actual analysis until it finishes running. This analysis will take about ___ minutes.</p>

<p>While it runs, we will set up a new analysis.</p>

<h3 id="changing-number-of-subsets">Changing Number of Subsets</h3>

<p>Open the Python tree in IcyTree. How many clades do you think are on this tree? What size are they? Make a copy of the pasta_script2 to pasta_script3. Add the option “–max-subproblem-size=”, and choose what you think is the maximum clade size on this tree. Edit the jobname to something unique.</p>

<h3 id="comparing-alignments">Comparing alignments</h3>

<p>As your alignments finish, copy them to your computer - the final alignments will be in the .aln file. We will view our alignment files in <a href="http://wasabiapp.org/">Wasabi</a>, which is a simple-browswer based alignment.</p>

<p>It can be very hard to appreciate the differences between alignments by eye. We will try making comparisons with <a href="https://github.com/smirarab/FastSP">FastSP</a>, which gives some at-a-glance comparisons.</p>

<p>Change directories back to your home directory. Clone the FastSP software:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cd ~/
git clone https://github.com/smirarab/FastSP.git
</code></pre></div></div>
<p>Now, change back to your previous directory with:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cd -
</code></pre></div></div>

<p>You can call FastSP like so:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>java -jar ~/FastSP/FastSP.jar -r reference_alignment_file -e estimated_alignment_file
</code></pre></div></div>

<p>So, for example, if I wanted to compare the alignment I estimated  with Pasta, using FastTree,  and the one with Pasta and Raxml, my command would look like</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>java -jar ~/FastSP/FastSP.jar -r /output/exercise_one/pastajob.marker001.pythonidae.aln -e /output/exercise_two/pastajob.marker001.pythonidae.aln
</code></pre></div></div>

<p>Based on these metrics, which alignment do you prefer? What other information would you want to know before choosing an alignment for your project?</p>

<p>##Coestimation of MSA and Phylogeny</p>

<p>Phylogenetic esitmation generally assumes that we know the alignment without error. We’ve already seen in our examples cases where we do not get the same alignment between methods. This is particularly true in areas that are hard to align. For example, in the below paper, they estimate a tree for fungi, which are deeply-diverged, and very diverse. The alignment has many problematic regions. What they showed in this paper was that by not accounting for the alignment uncertainty, support was overestimated for their tree hypothesis.</p>

<p><img src="../fig/lutzoni.png" alt="Lutzoni paper" /></p>

<p>We haven’t talked about Bayesian estimation, so I’m going to say very little on joint estimation of alignment and topology. This is a method that allows for a wide range of models to be deployed in order to estimate both the alignment and the tree, and shows great promise for difficult alignment issues. We have a few floating open labs throughout the semester, so if this is a topic of interest to people, we can revisit it then.</p>

<p>The primary software that performs this analysis is <a href="http://www.bali-phy.org/Tutorial3.html">Bali-Phy</a>, which is described <a href="http://www.bali-phy.org/Redelings_and_Suchard_2005.pdf">here</a>.</p>

<p>##References:</p>
<ul>
  <li>T-Coffee Manual: http://tcoffee.readthedocs.io/en/latest/</li>
  <li>Pasta Algorithm Description: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4424971/</li>
  <li>Sate Algorithm Description: https://academic.oup.com/sysbio/article/61/1/90/1680002</li>
</ul>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
  </ul>
</blockquote>





<div class="row">
  <div class="col-md-1">
    <h3>
      
      <a href="../13-Github/"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-md-10">
    
  </div>
  <div class="col-md-1">
    <h3>
      
      <a href="../"><span class="glyphicon glyphicon-menu-up" aria-hidden="true"></span><span class="sr-only">lesson home</span></a>
      
    </h3>
  </div>
</div>


      
      <footer>
  <div class="row">
    <div class="col-md-6" align="left">
      <h4>
	Copyright &copy; 2016
	
	<a href="http://datacarpentry.org">Data Carpentry</a>
	
      </h4>
    </div>
    <div class="col-md-6" align="right">
      <h4>
	<a href="/">Source</a>
	/
	<a href="/blob/gh-pages/CONTRIBUTING.md">Contributing</a>
	/
	<a href="/blob/gh-pages/CITATION">Cite</a>
	/
	<a href="">Contact</a>
      </h4>
    </div>
  </div>
</footer>

      
    </div>
  </body>
</html>
